### PR TITLE
Using cameras allows one viewport

### PR DESCRIPTION
**Pull request by Borkatator**
## About the PR 
This PR allows creating one viewport from the security cameras windows.

This viewport cannot be opened if another one exists, if you are not adjacend to the security computer, and closes automatically if you move away, or close the security camera window, in order to avoid exploits.

## Why's this needed? 
As a secoff, I sometimes like to watch cameras. However it takes the whole screen and I cannot see my colleagues anymore.
This PR would allow me to still see and talk to my colleagues while watching cameras, but hard limits on being able to open only 1 viewport, and not being able to move prevent this from being too powerful.

## Changelog

```changelog
(u)Borkatator
(*)You can now create a viewport window when watching camera feeds
```

### CODE REVIEW

**Comment by Superlagg** [comment]
Can you connect mechcomp things from the viewport to the main window with a multitool?

**Comment by Borkatator** [comment]
> Can you connect mechcomp things from the viewport to the main window with a multitool?

I'm not sure I understand the question.
Like using a multitool on a component in the viewport, and dragging to the main window to connect them together ?


I haven't tested this. I guess if this worked with previous viewports, it would still work.

**Comment by ZeWaka** [comment]
> Can you connect mechcomp things from the viewport to the main window with a multitool?

not if they use in_interact_range or similar....

**Comment by Superlagg** [comment]
> > Can you connect mechcomp things from the viewport to the main window with a multitool?
> 
> not if they use in_interact_range or similar....

Unless it's changed in the past year, you could use a camera to connect mechcomp things across the station. I know now it's limited in range, but i think it should still be possible if they're close enough.

**Comment by Borkatator** [comment]
> > > Can you connect mechcomp things from the viewport to the main window with a multitool?
> > 
> > 
> > not if they use in_interact_range or similar....
> 
> Unless it's changed in the past year, you could use a camera to connect mechcomp things across the station. I know now it's limited in range, but i think it should still be possible if they're close enough.


I just tested it, and yes, it tells me if they are more than 14 tiles apart, but aside from distance it doesn't prevent it.

I also pushed a new commit as I found a case where an Index error was raised.

File: code/obj/machinery/computer/security/security_cameras_chui.dm [hunk 1]
```diff
@@ -5,6 +5,24 @@ chui/window/security_cameras
 	windowSize = "650x500"
 	flags = CHUI_FLAG_MOVABLE | CHUI_FLAG_CLOSABLE
 
+	proc/create_viewport(client/target_client, turf/T)
+
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security/security_cameras_chui.dm:9]:
```suggestion
```

File: code/obj/machinery/computer/security/security_cameras_chui.dm [hunk 2]
```diff
@@ -5,6 +5,24 @@ chui/window/security_cameras
...
+		if(get_dist(owner,target_client.mob) > 1)
+			boutput(target_client,"<span class='alert'>You are too far to see the screen.</span>")
+		else
+			var/list/viewports = target_client.getViewportsByType("cameras: Viewport")
+			if(viewports.len >= 1)
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security/security_cameras_chui.dm:14]:
```suggestion
			if(length(viewports))
```

File: code/obj/machinery/computer/security.dm [hunk 3]
```diff
@@ -115,6 +140,7 @@
 	//winset(user, "camera_console.exitbutton", "command=\".windowclose \ref[src]\"")
 	//winshow(user, "camera_console", 1)
 
+	src.using_viewport = 0
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:143]:
```suggestion
	src.using_viewport = FALSE
```

File: code/obj/machinery/computer/security/security_cameras_chui.dm [hunk 4]
```diff
@@ -316,6 +337,15 @@ chui/window/security_cameras
 			if (istype(C))
 				owner.favorites -= C
 
+		else if (href_list["viewport"])
+			if (!owner.current)
+				boutput( clint, "<b>You need to select a camera before creating a viewport.</b>" )
+			else
+				owner.using_viewport = 1
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security/security_cameras_chui.dm:344]:
```suggestion
				owner.using_viewport = TRUE
```

File: code/obj/machinery/computer/security/security_cameras_chui.dm [hunk 5]
```diff
@@ -316,6 +337,15 @@ chui/window/security_cameras
 			if (istype(C))
 				owner.favorites -= C
 
+		else if (href_list["viewport"])
+			if (!owner.current)
+				boutput( clint, "<b>You need to select a camera before creating a viewport.</b>" )
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security/security_cameras_chui.dm:342]:
```suggestion
				boutput(clint, "<b>You need to select a camera before creating a viewport.</b>")
```

File: code/obj/machinery/computer/security.dm [hunk 6]
```diff
@@ -65,7 +77,20 @@
 			boutput(user, "<span class='alert'>ERROR. Cannot connect to camera.</span>")
 			playsound(src.loc, "sound/machines/buzz-sigh.ogg", 10, 0)
 			return
-		switchCamera(user, closest)
+		if (src.using_viewport)
+			if (length(clint.getViewportsByType("cameras: Viewport")) < 1)
+				src.using_viewport = 0
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:82]:
```suggestion
				src.using_viewport = FALSE
```

File: code/obj/machinery/computer/security.dm [hunk 7]
```diff
@@ -34,7 +45,8 @@
 		return 1
 
 	//moved out of global to only be used in sec computers
-	proc/move_security_camera(/*n,*/direct,var/mob/living/carbon/user)
+	proc/move_security_camera(/*n,*/direct,client/clint)
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:48]:
```suggestion
	proc/move_security_camera(direct, client/clint)
```

File: code/obj/machinery/computer/security.dm [hunk 8]
```diff
@@ -34,7 +45,8 @@
...
+		var/mob/living/carbon/user = clint.mob
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:49]: we can't assume this type

File: code/obj/machinery/computer/security.dm [hunk 9]
```diff
@@ -2,6 +2,7 @@
 	name = "Security Cameras"
 	icon_state = "security"
 	var/obj/machinery/camera/current = null
+	var/using_viewport = 0
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:5]:
```suggestion
	var/using_viewport = FALSE
```

File: code/obj/machinery/computer/security.dm [hunk 10]
```diff
@@ -20,6 +21,16 @@
 		..()
 		window = null
 
+	process()
+		..()
+		if(window)
+			for (var/client/subscriber in window.subscribers)
+				var/list/viewports = subscriber.getViewportsByType("cameras: Viewport")
+				if(get_dist(src,subscriber.mob) > 1 && viewports.len > 0)
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:29]:
```suggestion
				if(get_dist(src,subscriber.mob) > 1 && length(viewports))
```

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by Borkatator** [comment]
Thanks for the review.

I have also found that when using the viewport, selecting a camera changes the main view and not the viewport so i corrected this.

Since this caused a problem with the using_viewport flag, and found no good way to detect when the viewport window was closed, I ended up getting rid of the flag, and just checking the presence of a "cameras: Viewport" type viewport.

I have applied the other requested changes.

I have also tested it as a silicon, since I think that is what you meant by your comment:
`we can't assume this type`

File: code/obj/machinery/computer/security.dm [hunk 11]
```diff
@@ -33,8 +43,18 @@
...
+		var/turf/closestPos = null
+		for(var/i = 4, i >= 0 || !closestPos, i--)
+			closestPos = locate(T.x - i, T.y + i, T.z)
+			if(closestPos) break
+		vp.SetViewport(closestPos, 8, 8)
+
 	//moved out of global to only be used in sec computers
-	proc/move_security_camera(/*n,*/direct,var/mob/living/carbon/user)
+	proc/move_security_camera(direct, client/clint)
+		var/mob/user = clint.mob
```
Comment (ZeWaka) [inline code/obj/machinery/computer/security.dm:57]: Yeah I was just commenting on how you don't necessarily know that a human is interacting with it, you only know a mob is

**Review by ZeWaka — APPROVED** [review]

**Comment by Borkatator** [comment]
@ZeWaka sorry about that last rivet request, it's a misclick

**Pull request merged**